### DISCUSSION TITLE
running stateful services independently + scaling

### DISCUSSION THREAD

**Discussion by itsamemarkus** [Budibase/budibase#3364] [Help] [answered]
I don't have a use case for scaling yet, but I stumbled upon Budibase a few days ago and have been experimenting on a self hosted environment since. Very promising project. So many possibilities. 
(self hosted because I like having access to all data)

On the main readme you write

> Admin paradise. Budibase is made to scale. 

I was wondering if you could elaborate on this (maybe even in the official docs.)

I saw that you did a [great write up on k8s](https://docs.budibase.com/self-hosting/hosting-methods/kubernetes), maybe that's the future, but for testing purposes could you please explain how the services interact with each other?

For example a best practice tutorial on how to connect your own CouchDB  and S3 instance, because I'm not that familiar with the envoy proxy yet, so I can run the stateful services independently.

**Comment by shogunpurple** [comment, marked as answer]
Hi @itsamemarkus, great question.

You can read about how the different budibase services interact here: https://docs.budibase.com/references-1/budibase-architecture.

So the best way to run budibase right now in larger, high availability production grade deployments would be to use kubernetes due to the fact kubernetes makes it very easy to scale out services individually and provides some safeguards if a pod crashes, for example.

If using `docker-compose`, if you would like to connect to your own CouchDB cluster, S3 bucket or redis-cluster, you can update your existing `docker-compose.yaml` to point to those resources. The environment variables you need to update in `docker-compose.yaml` are `REDIS_URL`, `COUCHDB_URL` and `MINIO_URL`.  After updating these, you will need to run your `docker-compose --env-file hosting.properties up -d` command again for the changes to take effect.

Realistically, you should never have to update the `envoy.yaml` configuration for budibase - it is generally intended to run standalone and serves as a gateway for the budibase services, rather than something you should edit freely and use as your reverse proxy for other services other than budibase. You can of course do this if you wish by editing the `envoy.yaml` that comes with budibase - but generally we recommend against it.
